hypershift-operator/controllers/uwmtelemetry/uwm_telemetry.go

```diff
@@ -0,0 +1,376 @@
+		}
+		if !found {
+			continue
+		}
+		if url == telemetryRemoteWriteURL {
+			foundIndex = i
+			break
+		}
+	}
+
+	// remote write configuration comes from:
+	// https://github.com/openshift/cluster-monitoring-operator/blob/838a238342b2b1ab5c99a18bd271a7b15a1acbd1/pkg/manifests/manifests.go#L1665-L1720
+	telemetryRemoteWrite := RemoteWriteSpec{
+		URL: telemetryRemoteWriteURL,
+		Authorization: &monv1.SafeAuthorization{
+			Type: "Bearer",
+			Credentials: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: monitoring.UWMRemoteWriteSecret().Name,
+				},
+				Key: "token",
+			},
+		},
+		QueueConfig: &monv1.QueueConfig{
+			Capacity:          30000,
+			MaxSamplesPerSend: 10000,
+			BatchSendDeadline: "1m",
+			MinBackoff:        "1s",
+			MaxBackoff:        "256s",
+		},
+	}
+	if relabelConfig != nil {
+		telemetryRemoteWrite.WriteRelabelConfigs = []monv1.RelabelConfig{*relabelConfig}
+	}
+	telemetryRemoteWriteMap, err := toUnstructuredMap(telemetryRemoteWrite)
+	if err != nil {
+		return fmt.Errorf("cannot convert remote write config to unstructured map: %w", err)
+	}
+
+	if foundIndex != -1 {
+		remoteWriteConfigs[foundIndex] = telemetryRemoteWriteMap
+	} else {
+		remoteWriteConfigs = append(remoteWriteConfigs, telemetryRemoteWriteMap)
+	}
+	unstructured.SetNestedSlice(content, remoteWriteConfigs, "prometheus", "remoteWrite")
+	contentBytes, err := yaml.Marshal(content)
+	if err != nil {
+		return fmt.Errorf("cannot serialize configuration content: %w", err)
+	}
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	cm.Data["config.yaml"] = string(contentBytes)
+	return nil
+}
+
+func toUnstructuredMap(o interface{}) (map[string]interface{}, error) {
+	tempBytes, err := json.Marshal(o)
+	if err != nil {
+		return nil, err
+	}
+	result := map[string]interface{}{}
+	if err := json.Unmarshal(tempBytes, &result); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+var (
+	metricSelectorPattern = regexp.MustCompile(`__name__="([^"]+)"`)
+)
+
+func telemetryConfigToRelabelConfig(cm *corev1.ConfigMap) (*monv1.RelabelConfig, error) {
+	contentStr, ok := cm.Data["metrics.yaml"]
+	if !ok {
+		return nil, fmt.Errorf("telemetry config %s/%s does not include expected key: metrics.yaml", cm.Namespace, cm.Name)
+	}
+	content := map[string]interface{}{}
+	if err := yaml.Unmarshal([]byte(contentStr), &content); err != nil {
+		return nil, fmt.Errorf("unable to parse metrics selectors: %w", err)
+	}
+
+	selectors, exists, err := unstructured.NestedStringSlice(content, "matches")
+	if !exists || err != nil {
+		if !exists {
+			err = errors.New("'matches' not found")
+		}
+		return nil, fmt.Errorf("cannot extract selectors: %w", err)
+	}
+
+	var metricsToInclude []string
+	for _, sel := range selectors {
+		if match := metricSelectorPattern.FindStringSubmatch(sel); len(match) > 1 {
+			metricsToInclude = append(metricsToInclude, match[1])
+		}
+	}
+	return &monv1.RelabelConfig{
```

Comment:
what is this doing exactly?

Reply:
It is filtering the metrics sent to telemetry via remotewrite to only include those that are whitelisted for telemetry.
I am using the telemetry configuration configmap to determine this instead of us having to maintain/hardcode which 
metrics are suitable for telemetry.